src/devtools/client/debugger/src/components/Editor/LineHitCounts.tsx

```diff
@@ -1,52 +1,74 @@
 import { useMemo, useLayoutEffect, useState } from "react";
-import { interpolateLab } from "d3-interpolate";
-import { getLuminance } from "polished";
 import { useFeature } from "ui/hooks/settings";
 import { useAppDispatch, useAppSelector } from "ui/setup/hooks";
 
 import { setBreakpointHitCounts } from "../../actions/sources";
 import { HitCount, getHitCountsForSelectedSource } from "../../reducers/sources";
 import { getSelectedSourceId } from "../../selectors";
+import { getHitCountColors } from "../../utils/editor/hit-counts";
 
 import styles from "./LineHitCounts.module.css";
 
-type Props = { cm: any };
+type Props = { cm: any; editorRef: any };
 
 export default function LineHitCountsWrapper(props: Props) {
   const { value } = useFeature("hitCounts");
-  return value ? <LineHitCounts {...props} /> : null;
-}
-
-function LineHitCounts({ cm }: Props) {
-  const [isCollapsed, setIsCollapsed] = useState(true);
   const dispatch = useAppDispatch();
+
   const sourceId = useAppSelector(getSelectedSourceId);
   const hitCounts = useAppSelector(getHitCountsForSelectedSource);
   const hitCountMap = useMemo(
     () => (hitCounts !== null ? hitCountsToMap(hitCounts) : null),
     [hitCounts]
   );
-  const hitCountColorMap = useMemo(() => {
-    const uniqueHitCounts = Array.from(
-      new Set(hitCounts ? hitCounts.map(hitCount => hitCount.hits) : [])
-    );
-    const maxHitCount = Math.max(...uniqueHitCounts);
-
-    return new Map(
-      uniqueHitCounts.map(hitCount => [
-        hitCount,
-        Math.min(1, Math.max(0, Math.log(hitCount) / Math.log(maxHitCount))) || 0,
-      ])
-    );
-  }, [hitCounts]);
 
+  const hitCountColor = getHitCountColors(hitCounts);
   useLayoutEffect(() => {
     // HACK Make sure we load hit count metadata; normally this is done in response to a mouseover event.
     if (hitCounts === null) {
       dispatch(setBreakpointHitCounts(sourceId, 0));
     }
+
+    let styles = "";
+    for (const [
+      count,
+      { foregroundColor, backgroundColor, textColor, textHoverColor },
+    ] of hitCountColor) {
+      styles += `
+      .cm-s-mozilla .hit-count-${count} .CodeMirror-linenumber {
+        color: ${textColor};
+      }\n
+      .cm-s-mozilla .hit-count-${count}:hover .CodeMirror-linenumber {
+        color: ${textHoverColor};
+      }\n
+      
+      .hits-badge-${count} {
+        color: ${foregroundColor};
+        background-color: ${backgroundColor};
+      }\n
+      `;
```

Comment:
I'm concerned about this approach (if you're considering landing it).

Nested selectors can get pretty tricky as the number of style rules grow, and this approach to styling is outside of the half a dozen other approaches we're already using 😅

Reply:
I'm not sure how many options codemirror will give us. 

With that said, this is just me prototyping to see how this ui feels.

Reply:
Cool, cool 😁